interface/main/main_screen.php

```diff
@@ -15,14 +15,169 @@
+// Functions to support MFA.
+///////////////////////////////////////////////////////////////////////
 
-/* Include our required headers */
-require_once('../globals.php');
+function posted_to_hidden($name)
+{
+    if (isset($_POST[$name])) {
+        echo "<input type='hidden' name='" . attr($name) . "' value='" . attr($_POST[$name]) . "' />\n";
+    }
+}
+
+function generate_html_start($title)
+{
+    global $appId, $css_header;
+    ?>
+<html>
+<head>
+    <?php Header::setupHeader(); ?>
+<title><?php echo text($title); ?></title>
+<script src="<?php echo $GLOBALS['webroot'] ?>/library/js/u2f-api.js"></script>
+<script>
+function doAuth() {
+  var f = document.forms[0];
+  var requests = JSON.parse(f.form_requests.value);
+  // The server's getAuthenticateData() repeats the same challenge in all requests.
+  var challenge = requests[0].challenge;
+  var registeredKeys = new Array();
+  for (var i = 0; i < requests.length; ++i) {
+    registeredKeys[i] = {"version": requests[i].version, "keyHandle": requests[i].keyHandle};
+  }
+  u2f.sign(
+    '<?php echo addslashes($appId); ?>',
+    challenge,
+    registeredKeys,
+    function(data) {
+      if(data.errorCode && data.errorCode != 0) {
+        alert('<?php echo xla("Key access failed with error"); ?> ' + data.errorCode);
+        return;
+      }
+      f.form_response.value = JSON.stringify(data);
+      f.submit();
+    },
+    60
+  );
+}
+</script>
+</head>
+<body class='body_top'>
+<center>
+<h2><?php echo text($title); ?></h2>
+<form method="post"
+ action="main_screen.php?auth=login&site=<?php echo $_GET['site']; ?>"
+ target="_top" name="challenge_form">
+    <?php
+    posted_to_hidden('new_login_session_management');
+    posted_to_hidden('authProvider');
+    posted_to_hidden('languageChoice');
+    posted_to_hidden('authUser');
+    posted_to_hidden('clearPass');
+}
+
+function generate_html_end()
+{
+    echo "</form></center></body></html>\n";
+    session_unset();
+    session_destroy();
+    unset($_COOKIE[session_name()]);
+    return 0;
+}
+
+$errormsg = '';
+
+///////////////////////////////////////////////////////////////////////
+// Begin code to support U2F.
+///////////////////////////////////////////////////////////////////////
+
+$regs = array();          // for mapping device handles to their names
+$registrations = array(); // the array of stored registration objects
+$res1 = sqlStatement(
+    "SELECT a.name, a.var1 FROM login_mfa_registrations AS a " .
+    "WHERE a.user_id = ? AND a.method = 'U2F' ORDER BY a.name",
+    array($_SESSION['authId'])
+);
+while ($row1 = sqlFetchArray($res1)) {
+    $regobj = json_decode($row1['var1']);
+    $regs[json_encode($regobj->keyHandle)] = $row1['name'];
+    $registrations[] = $regobj;
+}
+if (!empty($registrations)) {
+    // There is at least one U2F key registered so we have to request or verify key data.
+    $scheme = isset($_SERVER['HTTPS']) ? "https://" : "http://";
+    $appId = $scheme . $_SERVER['HTTP_HOST'];
```

Comment:
Does this always work. For example, if a simple http server sitting behind a nginx reverse proxy that handles https? @sjpadgett , I think you have had to deal with this type of issue, correct?

Reply:
Actually U2F requires HTTPS.

Reply:
Reverse proxies can be a pain when dealing with protocols. I've found it best if you are calling a uri/url within the same domain, you are safest to use relative uri's and let server/browser resolve protocol. If calling outside domain then adding a protocol seems to be okay.

Reply:
@sunsetsystems, potential use case is a nginx docker functions as reverse proxy on the same machine and does all the https heavy lifting; and sends the traffic via http to another docker on same machine (btw, this is how the demo farm dockers work).

Reply:
The only use of $appId is to serve as a name that the u2f API requires to distinguish from other web applications. It's not used for network connections. But I'm not sure the browser API will be happy if it does not see a HTTPS connection. Needs testing I guess.

Reply:
oh, then shouldn't be an issue. The browser will definitely see HTTPS; just making sure you weren't relying on the server to tell you that.

Reply:
Let's plan to test this with docker, just in case.